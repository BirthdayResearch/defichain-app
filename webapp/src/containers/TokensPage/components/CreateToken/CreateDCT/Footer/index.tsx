import React from 'react';
import { I18n } from 'react-redux-i18n';
import { NavLink } from 'react-router-dom';
import classnames from 'classnames';
import { MdCheckCircle, MdErrorOutline } from 'react-icons/md';
import { Row, Col, Button } from 'reactstrap';

import styles from '../CreateDCT.module.scss';
import {
  TOKENS_PATH,
  MINIMUM_DFI_REQUIRED_FOR_TOKEN_CREATION,
} from '../../../../../../constants';
import { ITokenResponse } from '../../../../../../utils/interfaces';
import Spinner from '../../../../../../components/Svg/Spinner';
import { CreateTokenFormState } from '../..';
import ViewOnChain from '../../../../../../components/ViewOnChain';

interface CreateDCTProps {
  isUpdate: boolean;
  formState: CreateTokenFormState;
  isConfirmationModalOpen: string;
  setIsConfirmationModalOpen: (state: string) => void;
  cancelConfirmation: () => void;
  createConfirmation: () => void;
  updateConfirmation: () => void;
  wait: number;
  createdTokenData: ITokenResponse;
  isErrorCreatingToken: string;
  isErrorUpdatingToken: string;
  updatedTokenData: ITokenResponse;
  IsCollateralAddressValid: boolean;
}

const Footer: React.FunctionComponent<CreateDCTProps> = (
  props: CreateDCTProps
) => {
  const {
    isUpdate,
    formState,
    isConfirmationModalOpen,
    setIsConfirmationModalOpen,
    cancelConfirmation,
    createConfirmation,
    updateConfirmation,
    wait,
    createdTokenData,
    isErrorCreatingToken,
    updatedTokenData,
    isErrorUpdatingToken,
    IsCollateralAddressValid,
  } = props;

  return (
    <>
      <footer className='footer-bar'>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== 'default',
          })}
        >
          <Row className='justify-content-between align-items-center'>
            <Col className='col-auto'>
              <div className='caption-secondary'>
                {I18n.t('containers.tokens.createToken.dfiRequired')}
              </div>
              <div>
                {MINIMUM_DFI_REQUIRED_FOR_TOKEN_CREATION}
                &nbsp;
                {'DFI'}
              </div>
            </Col>
            <Col className='d-flex justify-content-end'>
              <Button
                to={TOKENS_PATH}
                tag={NavLink}
                color='link'
                className='mr-3'
              >
                {I18n.t('containers.tokens.createToken.cancel')}
              </Button>
              {isUpdate ? (
                <Button
                  disabled={
                    !formState.symbol ||
                    formState.symbol.length > 8 ||
                    formState.name.length > 128
                  }
                  onClick={() => {
                    setIsConfirmationModalOpen('confirm');
                  }}
                  color='primary'
                >
                  {I18n.t('containers.tokens.createToken.updateTokenButton')}
                </Button>
              ) : (
                <Button
                  disabled={
                    !formState.symbol ||
                    !formState.receiveAddress ||
                    formState.symbol.length > 8 ||
                    formState.name.length > 128 ||
                    !IsCollateralAddressValid
                  }
                  onClick={() => {
                    setIsConfirmationModalOpen('confirm');
                  }}
                  color='primary'
                >
                  {I18n.t('containers.tokens.createToken.createTokenButton')}
                </Button>
              )}
            </Col>
          </Row>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== 'confirm',
          })}
        >
          <div className='footer-sheet'>
            <dl className='row'>
              <dd className='col-12'>
                <span className='h2 mb-0'>
                  {isUpdate
                    ? I18n.t(
                        'containers.tokens.createToken.updateConfirmationText'
                      )
                    : I18n.t(
                        'containers.tokens.createToken.createConfirmationText'
                      )}
                </span>
              </dd>
            </dl>
          </div>
          <Row className='justify-content-between align-items-center'>
            <Col className='d-flex justify-content-end'>
              <Button
                color='link'
                className='mr-3'
                onClick={() => cancelConfirmation()}
              >
                {I18n.t('containers.masterNodes.createMasterNode.noButtonText')}
              </Button>
              <Button
                color='primary'
                onClick={() => {
                  isUpdate ? updateConfirmation() : createConfirmation();
                }}
                disabled={wait > 0 ? true : false}
              >
                {I18n.t(
                  'containers.masterNodes.createMasterNode.yesButtonText'
                )}
                &nbsp;
                <span className='timer'>{wait > 0 ? wait : ''}</span>
              </Button>
            </Col>
          </Row>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== 'loading',
          })}
        >
          <div className='footer-sheet'>
            <div className='text-center'>
              <Spinner />
            </div>
          </div>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== 'success',
          })}
        >
          <div className='footer-sheet'>
            <div className='text-center'>
              <p>
                {isUpdate
                  ? I18n.t('containers.tokens.createToken.updateTokenSuccess')
                  : I18n.t('containers.tokens.createToken.createTokenSuccess')}
              </p>
              <MdCheckCircle className='footer-sheet-icon' />
              <p>
                {`${I18n.t('containers.tokens.createToken.tokenHash')}: ${
                  isUpdate ? updatedTokenData.hash : createdTokenData.hash
                }`}
              </p>
            </div>
          </div>
          <Row className='justify-content-between align-items-center'>
            <Col className='d-flex justify-content-end'>
              <ViewOnChain
                txid={isUpdate ? updatedTokenData.hash : createdTokenData.hash}
              />
              <Button color='primary' to={TOKENS_PATH} tag={NavLink}>
                {I18n.t('containers.tokens.createToken.backToTokenPage')}
              </Button>
            </Col>
          </Row>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== 'failure',
          })}
        >
          <div className='footer-sheet'>
            <div className='text-center'>
              <MdErrorOutline
                className={classnames({
                  'footer-sheet-icon': true,
                  [styles[`error-dialog`]]: true,
                })}
              />
              <p>{isUpdate ? isErrorUpdatingToken : isErrorCreatingToken}</p>
            </div>
          </div>
          <div className='d-flex align-items-center justify-content-center'>
            <Button color='primary' to={TOKENS_PATH} tag={NavLink}>
              {I18n.t('containers.tokens.createToken.backToTokenPage')}
            </Button>
          </div>
        </div>
      </footer>
    </>
  );
};

export default Footer;
