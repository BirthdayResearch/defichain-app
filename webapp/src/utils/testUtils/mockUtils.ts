import PersistentStore from '../persistentStore';
import * as isElectron from '../isElectron';
import * as Notifications from '../notifications';
import log from 'loglevel';
import { runSaga } from 'redux-saga';
import axios from 'axios';

const mockPersistentStore = (getData, setData) => {
  PersistentStore.get = jest.fn(() => getData);
  PersistentStore.set = jest.fn(() => setData);
  return PersistentStore;
};

const mockAxios = mockFunc => {
  axios.create = jest.fn().mockImplementation(() => ({
    post: mockFunc,
  }));
};
const dispatchedFunc = async (method, payload = {}) => {
  const dispatched = Array();
  let action = {};
  if (payload) action = { payload };
  await runSaga(
    {
      dispatch: action => dispatched.push(action),
    },
    method,
    action
  );
  return dispatched;
};

const mockisElectron = () => {
  return jest.spyOn(isElectron, 'isElectron');
};
const mockIpcRenderer = () => {
  return jest.spyOn(isElectron, 'ipcRendererFunc');
};

const spyLogger = () => {
  return jest.spyOn(log, 'error');
};

const mockNotification = () => {
  return (
    jest
      .spyOn(Notifications, 'default')
      // tslint:disable-next-line: no-empty
      .mockImplementation(() => {})
  );
};
export {
  mockPersistentStore,
  mockAxios,
  dispatchedFunc,
  mockisElectron,
  mockIpcRenderer,
  spyLogger,
  mockNotification,
};
