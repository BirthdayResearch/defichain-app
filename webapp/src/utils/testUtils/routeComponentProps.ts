import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';
const history = createMemoryHistory();
const path = `/route/:id`;
const match: match<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);
delete history.location.key;
delete location.key;

export { history, match, location };
