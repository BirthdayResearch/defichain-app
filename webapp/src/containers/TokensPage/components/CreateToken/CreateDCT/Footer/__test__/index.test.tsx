import React from 'react';
import Footer from '../index';
import { shallow } from 'enzyme';
import * as routeComponentProps from '../../../../../../../utils/testUtils/routeComponentProps';

describe('Footer component', () => {
  it('should check for snapshot', () => {
    const props = {
      isUpdate: true,
      formState: {
        tokenInfo: [],
        fetchToken: () => {},
        createToken: () => {},
        updateToken: () => {},
        createdTokenData: {
          hash: '',
        },
        updatedTokenData: {
          hash: '',
        },
        isTokenUpdating: true,
        isErrorUpdatingToken: '',
        isTokenCreating: true,
        isErrorCreatingToken: '',
        paymentRequests: {
          label: '',
          id: '',
          time: '',
          address: '',
          message: '',
          amount: 1,
          unit: '',
          hdSeed: true,
        },
        receiveAddress: '',
        receiveLabel: '',
      },
      isConfirmationModalOpen: '',
      setIsConfirmationModalOpen: () => {},
      cancelConfirmation: () => {},
      createConfirmation: () => {},
      updateConfirmation: () => {},
      wait: 1,
      createdTokenData: {
        hash: '',
      },
      isErrorCreatingToken: '',
      isErrorUpdatingToken: '',
      updatedTokenData: {
        hash: '',
      },
      IsCollateralAddressValid: true,
    };

    const wrapper = shallow(<Footer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
