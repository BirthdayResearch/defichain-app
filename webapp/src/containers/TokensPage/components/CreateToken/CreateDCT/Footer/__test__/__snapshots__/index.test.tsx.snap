// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component should check for snapshot 1`] = `
<Fragment>
  <footer
    className="footer-bar"
  >
    <div
      className="d-none"
    >
      <Row
        className="justify-content-between align-items-center"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="col-auto"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <div
            className="caption-secondary"
          >
            DFI required
          </div>
          <div>
            101
             
            DFI
          </div>
        </Col>
        <Col
          className="d-flex justify-content-end"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Button
            className="mr-3"
            color="link"
            tag={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "propTypes": Object {
                  "activeClassName": [Function],
                  "activeStyle": [Function],
                  "aria-current": [Function],
                  "className": [Function],
                  "exact": [Function],
                  "innerRef": [Function],
                  "isActive": [Function],
                  "location": [Function],
                  "onClick": [Function],
                  "replace": [Function],
                  "sensitive": [Function],
                  "strict": [Function],
                  "style": [Function],
                  "target": [Function],
                  "to": [Function],
                },
                "render": [Function],
              }
            }
            to="/tokens"
          >
            Cancel
          </Button>
          <Button
            color="primary"
            disabled={true}
            onClick={[Function]}
            tag="button"
          >
            Update token
          </Button>
        </Col>
      </Row>
    </div>
    <div
      className="d-none"
    >
      <div
        className="footer-sheet"
      >
        <dl
          className="row"
        >
          <dd
            className="col-12"
          >
            <span
              className="h2 mb-0"
            >
              Are you sure you want to update a token?
            </span>
          </dd>
        </dl>
      </div>
      <Row
        className="justify-content-between align-items-center"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="d-flex justify-content-end"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Button
            className="mr-3"
            color="link"
            onClick={[Function]}
            tag="button"
          >
            Cancel
          </Button>
          <Button
            color="primary"
            disabled={true}
            onClick={[Function]}
            tag="button"
          >
            Continue
             
            <span
              className="timer"
            >
              1
            </span>
          </Button>
        </Col>
      </Row>
    </div>
    <div
      className="d-none"
    >
      <div
        className="footer-sheet"
      >
        <div
          className="text-center"
        >
          <Spinner />
        </div>
      </div>
    </div>
    <div
      className="d-none"
    >
      <div
        className="footer-sheet"
      >
        <div
          className="text-center"
        >
          <p>
            Token was updated successfully
          </p>
          <MdCheckCircle
            className="footer-sheet-icon"
          />
          <p>
            Token Hash: 
          </p>
        </div>
      </div>
      <Row
        className="justify-content-between align-items-center"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Col
          className="d-flex justify-content-end"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <ViewOnChain
            txid=""
          />
          <Button
            color="primary"
            tag={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "propTypes": Object {
                  "activeClassName": [Function],
                  "activeStyle": [Function],
                  "aria-current": [Function],
                  "className": [Function],
                  "exact": [Function],
                  "innerRef": [Function],
                  "isActive": [Function],
                  "location": [Function],
                  "onClick": [Function],
                  "replace": [Function],
                  "sensitive": [Function],
                  "strict": [Function],
                  "style": [Function],
                  "target": [Function],
                  "to": [Function],
                },
                "render": [Function],
              }
            }
            to="/tokens"
          >
            Back to token
          </Button>
        </Col>
      </Row>
    </div>
    <div
      className="d-none"
    >
      <div
        className="footer-sheet"
      >
        <div
          className="text-center"
        >
          <MdErrorOutline
            className="footer-sheet-icon undefined"
          />
          <p />
        </div>
      </div>
      <div
        className="d-flex align-items-center justify-content-center"
      >
        <Button
          color="primary"
          tag={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "exact": [Function],
                "innerRef": [Function],
                "isActive": [Function],
                "location": [Function],
                "onClick": [Function],
                "replace": [Function],
                "sensitive": [Function],
                "strict": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/tokens"
        >
          Back to token
        </Button>
      </div>
    </div>
  </footer>
</Fragment>
`;
